ext/mysql2/result.c

```diff
@@ -464,13 +488,17 @@ static VALUE rb_mysql_result_fetch_row_stmt(VALUE self, MYSQL_FIELD * fields, co
       }
     }
 
-    if (args->asArray) {
-      rb_ary_push(rowVal, val);
-    } else {
+    if (args->rowsAs == AS_HASH) {
       rb_hash_aset(rowVal, field, val);
+    } else /* array or struct */ {
+      rb_ary_push(rowVal, val);
     }
   }
 
+  if (args->rowsAs == AS_STRUCT) {
+    rowVal = cast_row_as_struct(self, rowVal, wrapper);
```

Comment:
This is still using an intermediary array in the row fetch loop above. Is that better/more efficient to create the struct once than to incrementally add members to the struct in the same way elements are pushed into the array or added into the hash? I think the latter would have more clarity / all three options being handled as cases in the same loop.

Reply:
I was trying to avoid modifying the the main loop for clarity! It also allowed me to factor the "cast_row_as_struct" code with a simple parameter list. The statement support added a lot of duplicate code and I really wanted to not contribute to that. Also, the struct.new native call has a nice array form.

Reply:
Happy to rethink this if you think it's critical.